### PR TITLE
Add TCP FastOpen support for macOS

### LINKED ISSUE #1632: Add TCP FastOpen support for macOS
## The `connectx` system call

This function is commonly used for inititiating a TCP Fast Open connection.

[Available since iOS 9 and OS X 10.11](https://en.wikipedia.org/wiki/TCP_Fast_Open)

```c
// Original definition in /usr/include/sys

__API_AVAILABLE(macosx(10.11), ios(9.0), tvos(9.0), watchos(2.0))
int connectx(int, const sa_endpoints_t *, sae_associd_t, unsigned int,
    const struct iovec *, unsigned int, size_t *, sae_connid_t *);

__API_AVAILABLE(macosx(10.11), ios(9.0), tvos(9.0), watchos(2.0))
int disconnectx(int, sae_associd_t, sae_connid_t);
```

Definition:

```c
int connectx(int socket,
             const sa_endpoints_t *endpoints,
             sae_associd_t associd /* Reserved, always be SAE_ASSOCID_ANY */,
             unsigned int flags,
             const struct iovec *iov,
             unsigned int iovcnt,
             size_t *len,
             sae_connid_t *connid /* Reserved, always be NULL */);
```

Example usage:

```c
struct sockaddr_in sa;
memcpy(&sa, /* TARGET ADDR */, sizeof(struct sockaddr_in));
sa.sin_len = sizeof(struct sockaddr_in);
sa_endpoints_t endpoints;
memset((char *)&endpoints, 0, sizeof(endpoints));
endpoints.sae_dstaddr    = (struct sockaddr *)&sa;
endpoints.sae_dstaddrlen = /* TARGET ADDR LEN */;

s = connectx(sockfd,
             &endpoints,
             SAE_ASSOCID_ANY,
             CONNECT_DATA_IDEMPOTENT | CONNECT_RESUME_ON_READ_WRITE,
             NULL,
             0,
             NULL,
             NULL);
```

### Variable definitions

- `SAE_ASSOCID_ANY`
    ```c
    // Defined in /usr/include/sys/socket.h (MacOSX.sdk)
    #define SAE_ASSOCID_ANY 0
    ```

- `CONNECT_DATA_IDEMPOTENT` and `CONNECT_RESUME_ON_READ_WRITE`
    ```c
    // Defined in /usr/include/sys/socket.h (MacOSX.sdk)
    #define CONNECT_RESUME_ON_READ_WRITE    0x1 /* resume connect() on read/write */
    #define CONNECT_DATA_IDEMPOTENT         0x2 /* data is idempotent */
    ```

### `sa_endpoints_t`

```c
// Defined in /usr/include/sys/socket.h (MacOSX.sdk)

/* sockaddr endpoints */
typedef struct sa_endpoints {
	unsigned int            sae_srcif;      /* optional source interface */
	const struct sockaddr   *sae_srcaddr;   /* optional source address */
	socklen_t               sae_srcaddrlen; /* size of source address */
	const struct sockaddr   *sae_dstaddr;   /* destination address */
	socklen_t               sae_dstaddrlen; /* size of destination address */
} sa_endpoints_t;
```

### `sae_associd_t`

```c
// Defined in /usr/include/sys/socket.h (MacOSX.sdk)

typedef __uint32_t sae_associd_t;
```

### PR DESCRIPTION
**Pull request by zonyitoo**
fix #1632

### CODE REVIEW

**Comment by zonyitoo** [comment]
@gnzlbg Hello, could you review this PR?

**Comment by gnzlbg** [comment]
@mati865 @pietroalbini there are many linker failures across the board of the form:

```
error: linking with `cc` failed: exit code: 1
  |
  = note: "cc" "-Wl,--as-needed" "-Wl,-z,noexecstack" "-L" "/rust/lib/rustlib/mips-unknown-linux-gnu/lib" "/checkout/target/mips-unknown-linux-gnu/debug/deps/linux_strerror_r-c9f2496a738ae509.15npztk0ht4i6q45.rcgu.o" "/checkout/target/mips-unknown-linux-gnu/debug/deps/linux_strerror_r-c9f2496a738ae509.1hrsp886jehe6mdx.rcgu.o" "/checkout/target/mips-unknown-linux-gnu/debug/deps/linux_strerror_r-c9f2496a738ae509.1ouetbkp3jjfhh6r.rcgu.o" "/checkout/target/mips-unknown-linux-gnu/debug/deps/linux_strerror_r-c9f2496a738ae509.1ssww7ta5gblmjyr.rcgu.o" "/checkout/target/mips-unknown-linux-gnu/debug/deps/linux_strerror_r-c9f2496a738ae509.1wvs1mat0k1jmlnr.rcgu.o" "/checkout/target/mips-unknown-linux-gnu/debug/deps/linux_strerror_r-c9f2496a738ae509.26fytuusgl87tw63.rcgu.o" "/checkout/target/mips-unknown-linux-gnu/debug/deps/linux_strerror_r-c9f2496a738ae509.2d0i2oja4i59c8oh.rcgu.o" "/checkout/target/mips-unknown-linux-gnu/debug/deps/linux_strerror_r-c9f2496a738ae509.2jigx46h41ggbo55.rcgu.o" "/checkout/target/mips-unknown-linux-gnu/debug/deps/linux_strerror_r-c9f2496a738ae509.2r2fv9qpzrw0s7gh.rcgu.o" "/checkout/target/mips-unknown-linux-gnu/debug/deps/linux_strerror_r-c9f2496a738ae509.2wcrd0gq8m7eh982.rcgu.o" "/checkout/target/mips-unknown-linux-gnu/debug/deps/linux_strerror_r-c9f2496a738ae509.2xa7qshjpomh3ts7.rcgu.o" "/checkout/target/mips-unknown-linux-gnu/debug/deps/linux_strerror_r-c9f2496a738ae509.31pxlxw0fjqhm0pg.rcgu.o" "/checkout/target/mips-unknown-linux-gnu/debug/deps/linux_strerror_r-c9f2496a738ae509.37mjsvnuqejb34tb.rcgu.o" "/checkout/target/mips-unknown-linux-gnu/debug/deps/linux_strerror_r-c9f2496a738ae509.398n1ac9a6g2l6mt.rcgu.o" "/checkout/target/mips-unknown-linux-gnu/debug/deps/linux_strerror_r-c9f2496a738ae509.3co2mfk5el1mjqjd.rcgu.o" "/checkout/target/mips-unknown-linux-gnu/debug/deps/linux_strerror_r-c9f2496a738ae509.3d17d7jg7smpu7ad.rcgu.o" "/checkout/target/mips-unknown-linux-gnu/debug/deps/linux_strerror_r-c9f2496a738ae509.3ezfcovkx4ijk7m3.rcgu.o" "/checkout/target/mips-unknown-linux-gnu/debug/deps/linux_strerror_r-c9f2496a738ae509.3hr6adqr8dmmonlo.rcgu.o" "/checkout/target/mips-unknown-linux-gnu/debug/deps/linux_strerror_r-c9f2496a738ae509.41swc9is3t913lxd.rcgu.o" "/checkout/target/mips-unknown-linux-gnu/debug/deps/linux_strerror_r-c9f2496a738ae509.46llah9hcnx2zwgu.rcgu.o" "/checkout/target/mips-unknown-linux-gnu/debug/deps/linux_strerror_r-c9f2496a738ae509.4g5zeiqv08xcngwe.rcgu.o" "/checkout/target/mips-unknown-linux-gnu/debug/deps/linux_strerror_r-c9f2496a738ae509.4j00w9sqr4uvy4zv.rcgu.o" "/checkout/target/mips-unknown-linux-gnu/debug/deps/linux_strerror_r-c9f2496a738ae509.4jwz2e02ltuj5wcn.rcgu.o" "/checkout/target/mips-unknown-linux-gnu/debug/deps/linux_strerror_r-c9f2496a738ae509.4no4j0q9dyfdgirf.rcgu.o" "/checkout/target/mips-unknown-linux-gnu/debug/deps/linux_strerror_r-c9f2496a738ae509.4oi9wcbvef3k2gbj.rcgu.o" "/checkout/target/mips-unknown-linux-gnu/debug/deps/linux_strerror_r-c9f2496a738ae509.4yqooiuh8050bntv.rcgu.o" "/checkout/target/mips-unknown-linux-gnu/debug/deps/linux_strerror_r-c9f2496a738ae509.5ecmxiouksc9p16v.rcgu.o" "/checkout/target/mips-unknown-linux-gnu/debug/deps/linux_strerror_r-c9f2496a738ae509.5gic5rctwbp57dke.rcgu.o" "/checkout/target/mips-unknown-linux-gnu/debug/deps/linux_strerror_r-c9f2496a738ae509.8loaqrsjy075q1l.rcgu.o" "/checkout/target/mips-unknown-linux-gnu/debug/deps/linux_strerror_r-c9f2496a738ae509.fppses3d1rky3g5.rcgu.o" "/checkout/target/mips-unknown-linux-gnu/debug/deps/linux_strerror_r-c9f2496a738ae509.gb2ijbqts8tapyg.rcgu.o" "/checkout/target/mips-unknown-linux-gnu/debug/deps/linux_strerror_r-c9f2496a738ae509.xed9cz7mimshle4.rcgu.o" "/checkout/target/mips-unknown-linux-gnu/debug/deps/linux_strerror_r-c9f2496a738ae509.zgga9h93wntxhsz.rcgu.o" "-o" "/checkout/target/mips-unknown-linux-gnu/debug/deps/linux_strerror_r-c9f2496a738ae509" "/checkout/target/mips-unknown-linux-gnu/debug/deps/linux_strerror_r-c9f2496a738ae509.rvlc6xnm1atdbpn.rcgu.o" "-Wl,--gc-sections" "-pie" "-Wl,-zrelro" "-Wl,-znow" "-nodefaultlibs" "-L" "/checkout/target/mips-unknown-linux-gnu/debug/deps" "-L" "/checkout/target/debug/deps" "-L" "/checkout/target/mips-unknown-linux-gnu/debug/build/libc-test-71fc6b33766ee9ae/out" "-L" "/checkout/target/mips-unknown-linux-gnu/debug/build/libc-test-71fc6b33766ee9ae/out" "-L" "/checkout/target/mips-unknown-linux-gnu/debug/build/libc-test-71fc6b33766ee9ae/out" "-L" "/checkout/target/mips-unknown-linux-gnu/debug/build/libc-test-71fc6b33766ee9ae/out" "-L" "/checkout/target/mips-unknown-linux-gnu/debug/build/libc-test-71fc6b33766ee9ae/out" "-L" "/checkout/target/mips-unknown-linux-gnu/debug/build/libc-test-71fc6b33766ee9ae/out" "-L" "/checkout/target/mips-unknown-linux-gnu/debug/build/libc-test-71fc6b33766ee9ae/out" "-L" "/checkout/target/mips-unknown-linux-gnu/debug/build/libc-test-71fc6b33766ee9ae/out" "-L" "/rust/lib/rustlib/mips-unknown-linux-gnu/lib" "-Wl,-Bstatic" "-Wl,--whole-archive" "-lcmsg" "-Wl,--no-whole-archive" "-Wl,--whole-archive" "-lerrqueue" "-Wl,--no-whole-archive" "-Wl,--whole-archive" "-lmain" "-Wl,--no-whole-archive" "-Wl,--whole-archive" "-llinux_strerror_r" "-Wl,--no-whole-archive" "-Wl,--whole-archive" "-llinux_fcntl" "-Wl,--no-whole-archive" "-Wl,--whole-archive" "-llinux_termios" "-Wl,--no-whole-archive" "-Wl,--whole-archive" "-llinux_ipv6" "-Wl,--no-whole-archive" "-Wl,--whole-archive" "-llinux_elf" "-Wl,--no-whole-archive" "/checkout/target/mips-unknown-linux-gnu/debug/deps/liblibc-91ebc08b121ef9ca.rlib" "-Wl,--start-group" "/rust/lib/rustlib/mips-unknown-linux-gnu/lib/libstd-861fd224266aa746.rlib" "/rust/lib/rustlib/mips-unknown-linux-gnu/lib/libpanic_unwind-799cf42449707bee.rlib" "/rust/lib/rustlib/mips-unknown-linux-gnu/lib/libhashbrown-b47cb6cf2bb91e6c.rlib" "/rust/lib/rustlib/mips-unknown-linux-gnu/lib/librustc_std_workspace_alloc-db287287628d70c2.rlib" "/rust/lib/rustlib/mips-unknown-linux-gnu/lib/libbacktrace-900a287ff1adfa65.rlib" "/rust/lib/rustlib/mips-unknown-linux-gnu/lib/libbacktrace_sys-0ae24a4b5240b9b6.rlib" "/rust/lib/rustlib/mips-unknown-linux-gnu/lib/librustc_demangle-06360ac9f1275b3d.rlib" "/rust/lib/rustlib/mips-unknown-linux-gnu/lib/libunwind-04674fd6529e87f1.rlib" "/rust/lib/rustlib/mips-unknown-linux-gnu/lib/libcfg_if-16719bfe5ee8d6c9.rlib" "/rust/lib/rustlib/mips-unknown-linux-gnu/lib/liblibc-a5dc8e1e92544cc6.rlib" "/rust/lib/rustlib/mips-unknown-linux-gnu/lib/liballoc-ae88b16fdea07518.rlib" "/rust/lib/rustlib/mips-unknown-linux-gnu/lib/librustc_std_workspace_core-2fa8c930ecf64b64.rlib" "/rust/lib/rustlib/mips-unknown-linux-gnu/lib/libcore-90f057afb960b3c3.rlib" "-Wl,--end-group" "/rust/lib/rustlib/mips-unknown-linux-gnu/lib/libcompiler_builtins-d82a8b0dd996842c.rlib" "-Wl,-Bdynamic" "-lc" "-lm" "-lrt" "-lpthread" "-lutil" "-ldl" "-lutil" "-ldl" "-lrt" "-lpthread" "-lgcc_s" "-lc" "-lm" "-lrt" "-lpthread" "-lutil" "-lutil"
  = note: /usr/bin/ld: /checkout/target/mips-unknown-linux-gnu/debug/deps/linux_strerror_r-c9f2496a738ae509.15npztk0ht4i6q45.rcgu.o: relocations in generic ELF (EM: 8)
          /usr/bin/ld: /checkout/target/mips-unknown-linux-gnu/debug/deps/linux_strerror_r-c9f2496a738ae509.15npztk0ht4i6q45.rcgu.o: relocations in generic ELF (EM: 8)
          /usr/bin/ld: /checkout/target/mips-unknown-linux-gnu/debug/deps/linux_strerror_r-c9f2496a738ae509.15npztk0ht4i6q45.rcgu.o: relocations in generic ELF (EM: 8)
          /usr/bin/ld: /checkout/target/mips-unknown-linux-gnu/debug/deps/linux_strerror_r-c9f2496a738ae509.15npztk0ht4i6q45.rcgu.o: relocations in generic ELF (EM: 8)
          /usr/bin/ld: /checkout/target/mips-unknown-linux-gnu/debug/deps/linux_strerror_r-c9f2496a738ae509.15npztk0ht4i6q45.rcgu.o: relocations in generic ELF (EM: 8)
          /usr/bin/ld: /checkout/target/mips-unknown-linux-gnu/debug/deps/linux_strerror_r-c9f2496a738ae509.15npztk0ht4i6q45.rcgu.o: relocations in generic ELF (EM: 8)
          /usr/bin/ld: /checkout/target/mips-unknown-linux-gnu/debug/deps/linux_strerror_r-c9f2496a738ae509.15npztk0ht4i6q45.rcgu.o: relocations in generic ELF (EM: 8)
          /usr/bin/ld: /checkout/target/mips-unknown-linux-gnu/debug/deps/linux_strerror_r-c9f2496a738ae509.15npztk0ht4i6q45.rcgu.o: relocations in generic ELF (EM: 8)
          /usr/bin/ld: /checkout/target/mips-unknown-linux-gnu/debug/deps/linux_strerror_r-c9f2496a738ae509.15npztk0ht4i6q45.rcgu.o: relocations in generic ELF (EM: 8)
          /usr/bin/ld: /checkout/target/mips-unknown-linux-gnu/debug/deps/linux_strerror_r-c9f2496a738ae509.15npztk0ht4i6q45.rcgu.o: relocations in generic ELF (EM: 8)
          /usr/bin/ld: /checkout/target/mips-unknown-linux-gnu/debug/deps/linux_strerror_r-c9f2496a738ae509.15npztk0ht4i6q45.rcgu.o: relocations in generic ELF (EM: 8)
          /usr/bin/ld: /checkout/target/mips-unknown-linux-gnu/debug/deps/linux_strerror_r-c9f2496a738ae509.15npztk0ht4i6q45.rcgu.o: relocations in generic ELF (EM: 8)
          /usr/bin/ld: /checkout/target/mips-unknown-linux-gnu/debug/deps/linux_strerror_r-c9f2496a738ae509.15npztk0ht4i6q45.rcgu.o: error adding symbols: file in wrong format
          collect2: error: ld returned 1 exit status
          

error: aborting due to previous error
```

Do these look familiar ?

**Comment by mati865** [comment]
I think there were changes to `cc` crate regarding relocations not long ago but I'll have to verify it.

**Comment by gnzlbg** [comment]
This one? Looks potentially related: https://github.com/alexcrichton/cc-rs/pull/461 cc @alexcrichton

**Comment by mati865** [comment]
It's not yet published so it was something else, Docker images shouldn't have changed AFAIK.

**Comment by mati865** [comment]
Wait, `/usr/bin/ld` is not the proper linker for cross compiled targets. Maybe some `cargo` change broke it?

**Comment by mati865** [comment]
I'll cc @ehuss ^

Possibly `cargo` or `cc` change made this no longer work:
https://github.com/rust-lang/libc/blob/74ff41223979af191278f53334fc81c645916d93/ci/docker/mips-unknown-linux-gnu/Dockerfile#L8

**Comment by ehuss** [comment]
Hm, sorry, that was an unexpected regression.  I'll work on a fix today, though it may take until sometime next week to get into nightly.

**Comment by gnzlbg** [comment]
@ehuss do you know which cargo PR caused it? I skimmed through the closed PR queue and couldn't find anything that ringed a bell.

**Comment by mati865** [comment]
I think that is https://github.com/rust-lang/cargo/pull/7649 and that's why I pointed to cargo (I knew about this PR before).

**Comment by ehuss** [comment]
Fix has been posted at rust-lang/cargo#7763.

**Comment by ehuss** [comment]
The fix should be in the latest nightly (2020-01-08).

**Comment by zonyitoo** [comment]
@gnzlbg  Could you trigger rebuild?

**Comment by zonyitoo** [comment]
@gnzlbg Hello, when could this PR be merged?

**Comment by JohnTitor** [comment]
Hey @zonyitoo, could you rebase to make sure things?

**Comment by zonyitoo** [comment]
Master branch haven't modified anything since I opened this PR.

**Comment by JohnTitor** [comment]
Style check complains about line length, could you fix it?

File: src/unix/bsd/apple/mod.rs [hunk 1]
```diff
@@ -3528,6 +3560,22 @@ extern "C" {
...
+        associd: sae_associd_t,
+        flags: ::c_uint,
+        iov: *const ::iovec,
+        iovcnt: ::c_uint,
+        len: *mut ::size_t,
+        connid: *mut sae_connid_t,
+    ) -> ::c_int;
+    pub fn disconnectx(
+        socket: ::c_int,
+        associd: sae_connid_t,
```
Comment (JohnTitor) [inline src/unix/bsd/apple/mod.rs]: It isn't `sae_associd_t`?
```suggestion
        associd: sae_associd_t,
```

Comment (zonyitoo) [reply src/unix/bsd/apple/mod.rs]: Ah! Yes!

**Pull request merged**